plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java

```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+                new CastTestCase("integer", "real"),
+                new CastTestCase("integer", "double"),
+                new CastTestCase("integer", "decimal(15)"),
+                new CastTestCase("integer", "decimal(10, 2)"),
+                new CastTestCase("integer", "decimal(30, 2)"),
+                new CastTestCase("smallint", "integer"),
+                new CastTestCase("char(10)", "char(501)"),
+                new CastTestCase("char(510)", "char(520)"),
+                new CastTestCase("char(510)", "varchar(1001)"),
+                new CastTestCase("char(10)", "varchar(1001)"),
+                new CastTestCase("char(10)", "varchar"),
+                new CastTestCase("varchar(10)", "varchar(1001)"),
+                new CastTestCase("varchar(10)", "varchar"),
+                new CastTestCase("varchar(1001)", "varchar(1010)"),
+                new CastTestCase("varchar(10)", "varbinary"),
+                new CastTestCase("timestamp", "date"),
+                new CastTestCase("date", "timestamp"),
+                new CastTestCase("date", "timestamp with time zone"))) {
+            try (TestTable table = new TestTable(getQueryRunner()::execute, "test_cast_", "(a %s)".formatted(testCase.fromType()))) {
+                assertThat(query("SELECT CAST(a AS %s) FROM %s".formatted(testCase.castType(), table.getName())))
+                        .isNotFullyPushedDown(ProjectNode.class);
+            }
+        }
+    }
+
+    @Test
+    public void testImplicitCastJoinPushdown()
+    {
+        String leftTable = "left_table_" + randomNameSuffix();
+        computeActual("CREATE TABLE %s (id int, varchar_50 varchar(50))".formatted(leftTable));
+        assertUpdate("insert into %s values (1, 'India')".formatted(leftTable), 1);
+        assertUpdate("insert into %s values (2, 'Poland')".formatted(leftTable), 1);
+
+        String rightTable = "right_table_" + randomNameSuffix();
+        computeActual("CREATE TABLE %s (varchar_100 varchar(100), capital varchar)".formatted(rightTable));
+        assertUpdate("insert into %s values ('India', 'New Delhi')".formatted(rightTable), 1);
+        assertUpdate("insert into %s values ('France', 'Paris')".formatted(rightTable), 1);
+
+        Session session = joinPushdownEnabled(getSession());
+
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "LEFT JOIN", "l.varchar_50 = r.varchar_100", Arrays.asList(1, 2));
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "RIGHT JOIN", "l.varchar_50 = r.varchar_100", Arrays.asList(1, null));
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "INNER JOIN", "l.varchar_50 = r.varchar_100", List.of(1));
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "INNER JOIN", "r.varchar_100 = l.varchar_50", List.of(1));
+
+        assertUpdate("DROP TABLE " + leftTable);
+        assertUpdate("DROP TABLE " + rightTable);
+    }
+
+    @Test
+    public void testExplicitCastJoinPushdown()
```

Comment:
I think it can be extracted to `TestCast` class and inject to `BaseOracleConnectorTest` as `@Nested`